vscode/src/completions/context/retrievers/recent-copy/recent-copy.ts

```diff
@@ -0,0 +1,106 @@
+import type { AutocompleteContextSnippet } from '@sourcegraph/cody-shared'
+import * as vscode from 'vscode'
+import type { ContextRetriever } from '../../../types'
+
+interface TrackedSelections {
+    timestamp: number
+    content: string
+    languageId: string
+    uri: vscode.Uri
+    startLine: number
+    endLine: number
+}
+
+export class RecentCopyRetriever implements vscode.Disposable, ContextRetriever {
+    // We use a map from the document URI to the set of tracked completions inside that document to
+    // improve performance of the `onDidChangeTextDocument` event handler.
+    private trackedSelections: TrackedSelections[] = new Array<TrackedSelections>()
+    public identifier = 'recent-copy'
+    private disposables: vscode.Disposable[] = []
+
+    constructor(
+        private readonly maxAgeMs: number,
+        private readonly maxSelections: number
+    ) {
+        this.disposables.push(
+            vscode.window.onDidChangeTextEditorSelection(this.onDidChangeTextEditorSelection.bind(this))
+        )
+    }
+
+    public async retrieve(): Promise<AutocompleteContextSnippet[]> {
+        const clipboardContent = await vscode.env.clipboard.readText()
+        const selectionItem = this.getSelectionItemIfExist(clipboardContent)
+        if (selectionItem) {
+            const autocompleteItem = {
+                uri: selectionItem.uri,
+                startLine: selectionItem.startLine,
+                endLine: selectionItem.endLine,
+                content: selectionItem.content,
+            }
+            return [autocompleteItem]
+        }
+        return []
+    }
+
+    public isSupportedForLanguageId(): boolean {
+        return true
+    }
+
+    private getSelectionItemIfExist(text: string): TrackedSelections | undefined {
+        for (const ts of this.trackedSelections) {
+            if (ts.content === text) {
+                return ts
+            }
+        }
+        return undefined
+    }
+
+    private async onDidChangeTextEditorSelection(
+        event: vscode.TextEditorSelectionChangeEvent
+    ): Promise<void> {
+        const editor = event.textEditor
+        this.addSelectionForTracking(editor.document, editor.selection)
+    }
```

Comment:
Do we need an async function here?

```suggestion
    private onDidChangeTextEditorSelection(event: vscode.TextEditorSelectionChangeEvent): void {
        const editor = event.textEditor
        this.addSelectionForTracking(editor.document, editor.selection)
    }
```

Reply:
oh right, we dont need async, removed it